std/range/package.d

```diff
@@ -6730,6 +6734,382 @@ if (isForwardRange!Source && hasLength!Source)
+        }
+        else
+        {
+            //Dollar token carries a static type, with no extra information.
+            //It can lazily transform into _source.length on algorithmic
+            //operations such as : pairs[$/2, $-1];
+            private static struct DollarToken
+            {
+                Slides!Source* mom;
+                @property size_t momLength()
+                {
+                    return mom.length;
+                }
+                alias momLength this;
+            }
+            DollarToken opDollar()
+            {
+                return DollarToken(&this);
+            }
+
+            //Slice overloads optimized for using dollar. Without this, to slice to end, we would...
+            //1. Evaluate pairs.length
+            //2. Multiply by _windowSize
+            //3. To finally just compare it (with min) to the original length of source (!)
+            //These overloads avoid that.
+            typeof(this) opSlice(DollarToken, DollarToken)
+            {
+                static if (hasSliceToEnd)
+                    return slides(_source[$ .. $], _windowSize);
+                else
+                {
+                    immutable len = _source.length;
+                    return slides(_source[len .. len], _windowSize);
+                }
+            }
+            typeof(this) opSlice(size_t lower, DollarToken)
+            {
+                import std.algorithm : min;
+                assert(lower <= length, "pairs slicing index out of bounds");
+                static if (hasSliceToEnd)
+                    return slides(_source[min(lower, _source.length) .. $], _windowSize);
+                else
+                {
+                    immutable len = _source.length;
+                    return slides(_source[min(lower, len) .. len], _windowSize);
+                }
+            }
+            typeof(this) opSlice(DollarToken, size_t upper)
+            {
+                assert(upper == length, "pairs slicing index out of bounds");
+                return this[$ .. $];
+            }
+        }
+    }
+
+    //Bidirectional range primitives
+    static if (hasSlicing!Source && hasLength!Source)
+    {
+        /**
+        Bidirectional range primitives. Provided only if both
+        $(D hasSlicing!Source) and $(D hasLength!Source) are $(D true).
+         */
+        @property auto back()
+        {
+            import std.algorithm : max;
+            assert(!empty, "back called on empty pairs");
+            immutable len = _source.length;
+            immutable start = max(0, len - _windowSize);
+            return _source[start .. $];
+        }
+
+        /// Ditto
+        void popBack()
+        {
+            assert(!empty, "popBack() called on empty pairs");
+            immutable end = (_source.length - 1) / _windowSize * _windowSize;
+            _source.popBack;
+            // update our "future" range too
+            _endSource.popBack;
+        }
+    }
+
+private:
```

Comment:
> Using Python's notation for private symbols with an underscore is useless with an explicit private feature

I know, but this style keeps popping up in phobos and the method that `slides` is based on (`chunks`) used the same notation.

> Struct members should be at the top

Again kept the style from `chunks`, but I do agree that it should be at top - updated :)